### DISCUSSION TITLE
Provide a guide how to get the original vertex data from boolean operations

### DISCUSSION THREAD

**Discussion by fire** [elalish/manifold#115] [Q&A]
Hi!

I was trying to get the original vertex data from boolean operations.

https://github.com/V-Sekai/godot/blob/csg-manifold-04/modules/csg/csg.cpp#L298-L419

Having trouble getting the original mesh id and its triangle index so I can check the property.

**Comment by elalish** [comment]
[`GetMeshRelation`](https://elalish.github.io/manifold/classmanifold_1_1_manifold.html#a45ea90568e3c9072089e9854cd51a02f) gives you the mesh id and triangles indices, as well as barycentric coordinates. `GetMeshIDs` gets the mesh ids. For any newly-constructed manifold this vector will length 1 (since it's a new meshID). If you can give some specifics of the trouble you're having, I'll be happy to help more.

**Comment by elalish** [comment]
These test functions might also give you some idea of how it's used: https://github.com/elalish/manifold/blob/master/test/mesh_test.cpp#L36-L79

**Comment by fire** [comment]
Can you do an example of getting the triangle from mesh_b when you boolean intersect a and b? Or boolean add if it's too hard.

**Comment by elalish** [comment]
Is this what you're after?
```
Mesh a_mesh = ImportMesh('something.glb');
Manifold a(a_mesh);
Manifold b = Manifold.Cube();
Mesh b_mesh = b.GetMesh();
const int aID = a.GetMeshIDs()[0];
const int bID = b.GetMeshIDs()[0];

Manifold c = a ^ b;
MeshRelation rel = c.GetMeshRelation();
Mesh c_mesh = c.GetMesh();
std::vector<int> meshID2Original = Manifold::MeshID2Original();

int c_tri = 0; // referring to c_mesh.triVerts[c_tri]
int triVert = 2; // 0, 1, or 2
BaryRef ref = rel.triBary[c_tri];
Mesh& inMesh = meshID2Original[ref.meshID] == aID ? a_mesh : b_mesh;
glm::ivec3 inTri = inMesh.triVerts[ref.tri];
glm::vec3 uvw = rel.UVW(c_tri, triVert); 

glm::vec3 inProp = {inMesh.vertProp[inMesh.triVerts[inTri][0]],
                            inMesh.vertProp[inMesh.triVerts[inTri][1]],
                            inMesh.vertProp[inMesh.triVerts[inTri][2]]};
float prop = glm::dot(inProp, uvw);

// Or check if(ref.vertBary[triVert] < 0) then this refers to c_mesh.vertPos[ref.vertBary[triVert] + 3]. 
// Otherwise it's a new vertex that will have different properties for each side of the intersection.
```
This same system works even after many consecutive boolean operations, since it will point each triangle back to whichever original mesh it came from, not just the most recent inputs.

**Comment by fire** [comment]
Thanks. Maybe we can add to the documentation or a test, but I was able to use this.
